### PR TITLE
CREATE/ALTER ROLE RESOURCE GROUP

### PR DESCRIPTION
**Pull request by pivotal-ning-yu**
In this patchset we implemented the CREATE/ALTER ROLE RESOURCE GROUP syntax, testcases were also provided.

### CODE REVIEW

**Comment by pivotal-ning-yu** [comment]
@xiong-gang @yydzero @eurekaka @guofengrichard @pengzhout Please help to review this PR, thanks in advance.

**Comment by pivotal-ning-yu** [comment]
Updated one of the patch, removed the unrelated changes to resscheduler.c

**Comment by pivotal-ning-yu** [comment]
BTW, help information like below are not implemented yet as they are part of documentation which needs modification on `doc/src/sgml/ref/{create/alter}_role.sgml`.

\h CREATE ROLE
\h ALTER ROLE

File: src/backend/commands/user.c [hunk 1]
```diff
@@ -495,11 +508,51 @@ CreateRole(CreateRoleStmt *stmt)
...
+		rsgid = GetResGroupIdForName(resgroup, ShareLock);
+		if (rsgid == InvalidOid)
+			ereport(ERROR,
+					(errcode(ERRCODE_UNDEFINED_OBJECT),
+					 errmsg("resource group \"%s\" does not exist",
+						 resgroup)));
+		else if (rsgid == ADMINRESGROUP_OID && !issuper)
+			ereport(ERROR,
+					(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
+					 errmsg("only super user can be assigned to admin resgroup")));
```
Comment (danielgustafsson) [inline src/backend/commands/user.c]: s/super user/superuser/

File: src/backend/commands/user.c [hunk 2]
```diff
@@ -763,6 +818,15 @@ AlterRole(AlterRoleStmt *stmt)
...
 			if (1 == numopts) alter_subtype = "RESOURCE QUEUE";
 		}
+		else if (strcmp(defel->defname, "resourceGroup") == 0)
+		{
+			if (dresgroup)
+				ereport(ERROR,
+						(errcode(ERRCODE_SYNTAX_ERROR),
+						 errmsg("conflicting or redundant options")));
+			dresgroup = defel;
+			if (1 == numopts) alter_subtype = "RESOURCE GROUP";
```
Comment (danielgustafsson) [inline src/backend/commands/user.c]: Break into two lines.

File: src/backend/commands/user.c [hunk 3]
```diff
@@ -1064,6 +1130,56 @@ AlterRole(AlterRoleStmt *stmt)
...
+
+			if (Gp_role == GP_ROLE_DISPATCH)
+				ereport(NOTICE,
+						(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						 errmsg("resource group required -- "
+								"using default resource group \"%s\"",
+								resgroup)));
+		}
+
+		if (!strcmp(resgroup, "none"))
```
Comment (danielgustafsson) [inline src/backend/commands/user.c]: Why having two if statements with the same test?

File: src/backend/commands/user.c [hunk 4]
```diff
@@ -1064,6 +1130,56 @@ AlterRole(AlterRoleStmt *stmt)
...
+			rsgid = GetResGroupIdForName(resgroup, ShareLock);
+			if (rsgid == InvalidOid)
+				ereport(ERROR,
+					(errcode(ERRCODE_UNDEFINED_OBJECT),
+					 errmsg("resource group \"%s\" does not exist",
+							resgroup)));
+			else if (rsgid == ADMINRESGROUP_OID && !bWas_super)
+				ereport(ERROR,
+						(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
+						 errmsg("only super user can be assigned to admin resgroup")));
```
Comment (danielgustafsson) [inline src/backend/commands/user.c]: s/super user/superuser/

**Comment by pivotal-ning-yu** [comment]
@danielgustafsson Thanks very much for the review, one commit is submitted to correct the issues pointed by you.

I also rebased the patchset to master to resolve some conflicts.

File: src/backend/commands/user.c [hunk 5]
```diff
@@ -288,6 +291,14 @@ CreateRole(CreateRoleStmt *stmt)
 						 errmsg("conflicting or redundant options")));
 			dresqueue = defel;
 		}
+		else if (strcmp(defel->defname, "resourceGroup") == 0)
+		{
+			if (dresgroup)
+				ereport(ERROR,
+						(errcode(ERRCODE_SYNTAX_ERROR),
+						 errmsg("conflicting or redundant options")));
```
Comment (xiong-gang) [inline src/backend/commands/user.c]: The error message can be more specific here.

**Comment by xiong-gang** [comment]
LGTM

File: src/backend/commands/user.c (hunk 5 above)
Comment (pivotal-ning-yu) [inline src/backend/commands/user.c]: @xiong-gang thanks for the review. I checked the context of this code, other options are also producing the same short error message, e.g.:

```C
        else if (strcmp(defel->defname, "superuser") == 0)
        {
            if (dissuper)
                ereport(ERROR,
                        (errcode(ERRCODE_SYNTAX_ERROR),
                         errmsg("conflicting or redundant options")));
            dissuper = defel;
        }
```

So how about improve all of them later in a separate PR?

**Comment by pivotal-ning-yu** [comment]
We rebased to origin/master to resolve the conflicts.

**Comment by pivotal-ning-yu** [comment]
After a rebase there is no conflict in code and all CI tests passed, per our discussion we will merge it now.

**Pull request merged**